### PR TITLE
Change gpflow2 Scipy optimizer default method to L-BFGS-B (as in gpflow1)

### PR DESCRIPTION
**Pull request by st--**
Explicitly sets the default optimization method for Scipy() to L-BFGS-B (as in gpflow1), instead of letting scipy.minimize() figure it out. This resulted in BFGS, which may be too aggressive when encountering complex optimization surfaces, so L-BFGS-B may be a more sensible default.

### CODE REVIEW

**Comment by awav** [comment]
Are we sure that we want this change? I would say that default value shoudl bfgs, we never discussed why gpflow scipy version had l-bfgs by default.

**Comment by awav** [comment]
In current scipy.minimize method we could just reference to the documentation in https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html, after that change we will have to make a note that our default value is different.

**Comment by st--** [comment]
`BFGS` estimates the full Hessian at each step, which makes it super slow for models with a reasonably large number of parameters, as it constructs a #param x #param matrix (also gives a big memory hit). This is what resulted in the significant slow-down in the Coregionalisation notebook (#1072, #1090). Optimising models with 10^2 inducing points -> 10^4 parameters is fairly common. Operating with 10^4 x 10^4 matrices by default is painful. I think `L-BFGS-B` is a more sensible default (and reproduces gpflow1 behaviour). I do agree that we should carefully think about this - e.g. it makes lots of tests fail (though I also wonder how they passed before in gpflow1!), which is why I split it off from #1111...

**Comment by markvdw** [comment]
I'm with Ti.

File: gpflow/optimizers/scipy.py [hunk 1]
```diff
@@ -14,6 +14,8 @@
 
 
 class Scipy:
+    _DEFAULT_METHOD = 'L-BFGS-B'
```
Comment (awav) [inline gpflow/optimizers/scipy.py]:
```suggestion
```

File: gpflow/optimizers/scipy.py [hunk 2]
```diff
@@ -14,6 +14,8 @@
...
+
```
Comment (awav) [inline gpflow/optimizers/scipy.py]:
```suggestion
```

File: gpflow/optimizers/scipy.py [hunk 3]
```diff
@@ -14,6 +14,8 @@
...
     def minimize(self,
                  closure: LossClosure,
                  variables: Variables,
```
Comment (awav) [inline gpflow/optimizers/scipy.py]:
```suggestion
                  variables: Variables,
                  method: Optional[str] = "L-BFGS-B",
```

File: gpflow/optimizers/scipy.py [hunk 4]
```diff
@@ -46,14 +48,18 @@ def minimize(self,
...
         func = self.eval_func(closure, variables)
         if step_callback is not None:
             if 'callback' in scipy_kwargs:
                 raise ValueError("Callback passed both via `step_callback` and `callback`")
 
             callback = self.callback_func(closure, variables, step_callback)
             scipy_kwargs.update(dict(callback=callback))
-        return scipy.optimize.minimize(func, initial_params, jac=True, **scipy_kwargs)
+
+        method = scipy_kwargs.pop('method', self._DEFAULT_METHOD)
```
Comment (awav) [inline gpflow/optimizers/scipy.py]:
```suggestion
```

**Comment by st--** [comment]
Sure! I only did it this way because that's how it was in the old (gpflow1) `ScipyOptimizer`, so I thought there was a good reason to do it that way...

**Pull request merged**